nvdaHelper/remote/excel.cpp

```diff
@@ -0,0 +1,472 @@
+	SelectObject(tempDC, hOldFont);
+	DeleteObject(hFont);
+	SelectObject(tempDC, hOldBmp);
+	DeleteObject(hBmp);
+	DeleteDC(tempDC);
+	return sizl.cx;
+}
+
+__int64 getCellStates(HWND hwnd, IDispatch* pDispatchRange) {
+	__int64 states=0;
+	// If the current row is a summary row, expose the collapsed or expanded states depending on wither the inner rows are showing or not.
+	CComPtr<IDispatch> pDispatchRow=nullptr;
+	HRESULT res=_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_ENTIREROW,VT_DISPATCH,&pDispatchRow);
+	if(FAILED(res)) {
+		LOG_DEBUGWARNING(L"range.entireRow failed with code "<<res);
+	}
+	if(pDispatchRow) {
+		BOOL summary=false;
+		res=_com_dispatch_raw_propget(pDispatchRow,XLDISPID_ROW_SUMMARY,VT_BOOL,&summary);
+		if(FAILED(res)) {
+			LOG_DEBUGWARNING(L"row.summary failed with code "<<res);
+		}
+		if(summary) {
+			BOOL showDetail=false;
+			res=_com_dispatch_raw_propget(pDispatchRow,XLDISPID_ROW_SHOWDETAIL,VT_BOOL,&showDetail);
+			if(FAILED(res)) {
+				LOG_DEBUGWARNING(L"row.showDetail failed with code "<<res);
+			}
+			states+=(showDetail?NVSTATE_EXPANDED:NVSTATE_COLLAPSED);
+		}
+	}
+	// If this row was neither collapsed or expanded, then try the same for columns instead. 
+	if(!(states&NVSTATE_EXPANDED)&&!(states&NVSTATE_COLLAPSED)) {
+		CComPtr<IDispatch> pDispatchColumn=nullptr;
+		res=_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_ENTIRECOLUMN,VT_DISPATCH,&pDispatchColumn);
+		if(FAILED(res)) {
+			LOG_DEBUGWARNING(L"range.entireColumn failed with code "<<res);
+		}
+		if(pDispatchColumn) {
+			BOOL summary=false;
+			res=_com_dispatch_raw_propget(pDispatchColumn,XLDISPID_COLUMN_SUMMARY,VT_BOOL,&summary);
+			if(FAILED(res)) {
+				LOG_DEBUGWARNING(L"column.summary failed with code "<<res);
+			}
+			if(summary) {
+				BOOL showDetail=false;
+				res=_com_dispatch_raw_propget(pDispatchColumn,XLDISPID_COLUMN_SHOWDETAIL,VT_BOOL,&showDetail);
+				if(FAILED(res)) {
+					LOG_DEBUGWARNING(L"column.showDetail failed with code "<<res);
+				}
+				states+=(showDetail?NVSTATE_EXPANDED:NVSTATE_COLLAPSED);
+			}
+		}
+	}
+	// Expose whether this cell has a formula
+	BOOL hasFormula=false;
+	res=_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_HASFORMULA,VT_BOOL,&hasFormula);
+	if(FAILED(res)) {
+		LOG_DEBUGWARNING(L"range.hasFormula failed with code "<<res);
+	}
+	if(hasFormula) {
+		states+=NVSTATE_HASFORMULA;
+	}
+	// Expose whether this cell has a dropdown menu for choosing valid values
+	CComPtr<IDispatch> pDispatchValidation=nullptr;
+	res=_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_VALIDATION,VT_DISPATCH,&pDispatchValidation);
+	if(FAILED(res)) {
+		LOG_DEBUGWARNING(L"range.validation failed with code "<<res);
+	}
+	if(pDispatchValidation) {
+		long validationType=0;
+		res=_com_dispatch_raw_propget(pDispatchValidation,XLDISPID_VALIDATION_TYPE,VT_I4,&validationType);
+		if(FAILED(res)) {
+			//LOG_DEBUGWARNING(L"validation.type failed with code "<<res);
```

Comment:
why is this commented out?

Reply:
It always produces an error. I have now suppressed this log call only for that particular error, and commented why.